### PR TITLE
New hosting model

### LINKED ISSUE #54: [ENH] Update bundled .NET Runtime to version 5.0.0 RC2

### LINKED ISSUE #55: [ENH] New hosting model
# Overview 

Right now GmodDotNet uses `hostfxr_initialize_for_runtime_config`-centered hosting API. This was the only API available at .NET Core version 3.1 and is suboptimal for GmodDotNet application model: the actual main assembly load context of the application was not the default one, which makes shared dependencies resolution process complicated and runtime code base hard to maintain properly.

.NET 5 has introduced a new hosting model based on `hostfxr_initialize_for_dotnet_command_line` and `hdt_get_function_pointer`. 

# Enhancements which will come with the new hosting model 

1) It will be easier to maintain hosting layer codebase.

2) More predictable dependencies resolution process.

3) There will be no `GmodNET.API` types duplications.

### PR DESCRIPTION
**Pull request by GlebChili**
Closes issues #54 and #55.

### CODE REVIEW

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs
Comment (GlebChili) [inline gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: @AaronJThompson As I mentioned in previous PR, address-of operator now works properly with functions with unmanaged calling conventions.

File: gm_dotnet_managed/GmodNET/Startup.cs
Comment (AaronJThompson) [inline gm_dotnet_managed/GmodNET/Startup.cs:14]: Was this deprecated/unused? Or has it been replaced with something?

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs
Comment (AaronJThompson) [reply gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: Nice! Looks much better

**Review by AaronJThompson — COMMENTED** [review]
The native re-write looks far nicer, great work! As I've said before, my C++ experience is very minimal, however, this looks much nicer. The only thing I can suggest from a readability point of view is maybe refactoring the `reinterpret_cast`'s. I'm submitting this review without approval for now due to the previous comments, however, I will approve if they do not apply/don't make sense. Great work otherwise!

File: gm_dotnet_managed/GmodNET/Startup.cs
Comment (AaronJThompson) [reply gm_dotnet_managed/GmodNET/Startup.cs:14]: From the C++ code, am I right in saying we just call Main directly now instead?

Comment (GlebChili) [reply gm_dotnet_managed/GmodNET/Startup.cs:14]: You are right: previous versions of hostfxr API required a delegate type declared to make a call to managed function. It was needed for various reasons, like specifying native calling conventions. With .NET 5 and it's new `UnmanagedCallersOnly` attribute we no longer need this declaring delegates at all.

**Comment by GlebChili** [comment]
@AaronJThompson Let's wait for @phaseflowcentral to have a look at C++ code. He has a formal education in modern C++, so he knows better 😊.

**Comment by GlebChili** [comment]
@AaronJThompson Unfortunately @phaseflowcentral is too busy for now. What was your concern with `reinterpret_cast`?

**Comment by AaronJThompson** [comment]
> @AaronJThompson Unfortunately @phaseflowcentral is too busy for now. What was your concern with `reinterpret_cast`?

It's fine I'll approve. I don't have much experience with C++ so I'm not sure if it's possible to refactor it

**Review by AaronJThompson — APPROVED** [review]
Approved. Greate job!

**Pull request merged**
